### ISSUE TITLE
Bug: no-loop-func is complaining about global variables modified outside of the loop

### ISSUE THREAD

**Issue by greggman** [eslint/eslint#17793]
### Environment

Node version: N/A
npm version: N/A
Local ESLint version: v8.54.0
Global ESLint version: v8.54.0
Operating System: N/A

Running in the playground ESLint: v8.54.0

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "env": {
        "es6": true,
        "browser": true
    },
    "parserOptions": {
        "ecmaFeatures": {},
        "ecmaVersion": "latest",
        "sourceType": "script"
    },
    "rules": {
        "no-loop-func": [
            "error"
        ]
    }
}
```
</details>

```js
let data;
(function() {
  data = {time: new Date.now()};
})();

const bar = [];

for (let i = 0; i < 3; ++i) {
  const foo = () => {
    console.log(data?.time);
  };
  bar.push(foo);
}
```

### What did you expect to happen?

eslint doesn't complain about accessing `data`

### What actually happened?

9:15

Function declared in a loop contains unsafe references to variable(s) 'data'.  ([no-loop-func](https://eslint.org/docs/rules/no-loop-func))

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoibGV0IGRhdGE7XG4oZnVuY3Rpb24oKSB7XG4gIGRhdGEgPSB7dGltZTogbmV3IERhdGUubm93KCl9O1xufSkoKTtcblxuY29uc3QgYmFyID0gW107XG5cbmZvciAobGV0IGkgPSAwOyBpIDwgMzsgKytpKSB7XG4gIGNvbnN0IGZvbyA9ICgpID0+IHtcbiAgICBjb25zb2xlLmxvZyhkYXRhPy50aW1lKTtcbiAgfTtcbiAgYmFyLnB1c2goZm9vKTtcbn1cbiIsIm9wdGlvbnMiOnsiZW52Ijp7ImVzNiI6dHJ1ZSwiYnJvd3NlciI6dHJ1ZX0sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In0sInJ1bGVzIjp7Im5vLWxvb3AtZnVuYyI6WyJlcnJvciJdfX19

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Is this behaving as expected or is it a bug?

To me, `data` is a global variable and the `no-loop-func` should not be complaining here. If I change `data` to `const data = {time: 123}` and delete the function setting it the issue goes away.

Of course the actual place this happened is more complicated. I can comment out the rule for the line the error is happening but it didn't seem like it should be an error.

**Comment by mdjermanovic** [comment]
This rule warns if a variable that's scoped outside of the loop and used in a function inside the loop is modified in or after the loop. When a variable is modified in an inner function, the rule can't reliably know when that function is called so it considers that the function might be called (and thus the variable might be modified) in or after the loop.

In your example though, the variable is obviously modified only before the loop because the assignment is in an IIFE that's executed before the loop, so the code is logically the same as:

```js
let data;
data = {time: new Date.now()};

const bar = [];

for (let i = 0; i < 3; ++i) {
  const foo = () => {
    console.log(data?.time);
  };
  bar.push(foo);
}

```

for which this rule doesn't report errors.

Would it solve the problem if we update the rule to better handle IIFEs?

**Comment by greggman** [comment]
I'm going to close this. I think the current check is valid as is. For the specific example I can fix it with

```js
const data = {time: new Date.now()};

const bar = [];

for (let i = 0; i < 3; ++i) {
  const foo = () => {
    console.log(data?.time);
  };
  bar.push(foo);
}
```

And I think in my actual code that prompted this report the warning is correct and I should fix my code.
